activerecord/lib/active_record/persistence.rb

```diff
@@ -526,7 +526,12 @@ def composite_query_constraints_list # :nodoc:
       #   todos = [1,2,3]
       #   Todo.destroy(todos)
       def destroy(id)
-        if id.is_a?(Array)
+        multiple_ids = id.is_a?(Array) && (
```

Comment:
It looks like the `composite_primary_key?` can be extracted and only done once like:

```ruby
        multiple_ids = if composite_primary_key?
             id.first.is_a?(Array)
        else
             id.is_a?(Array)
         end
```
Which can be inlined 

```ruby
(composite_primary_key? && id.first.is_a?(Array)) || id.is_a?(Array)
```

It seems better for readability but technically may perform worse as we will always perform at least two checks while in the proposed implementation in most case the condition will stop at `id.is_a?(Array)`

Though such performance difference will only be visible on a micro benchmark and unlikely noticeable in scope of a full unit of work. Also since `composite_primary_key?` is expected to almost never change in runtime the code may end up being optimized by a JIT compiler

Reply:
Let's do the non-inlined suggestion. I think the current code is too hard to read and it took me awhile to parse what it was doing.